File: examples/python/inventory/inventory_logic.py
"""Inventory command handlers.

Business logic for processing inventory commands.
"""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains

from state import InventoryState


class CommandRejectedError(Exception):
    """Command was rejected due to business rule violation."""


def handle_initialize_stock(command_book, command_any, state: InventoryState, seq: int, log) -> angzarr.EventBook:
    if state.exists():
        raise CommandRejectedError("Inventory already initialized")

    cmd = domains.InitializeStock()
    command_any.Unpack(cmd)

    if not cmd.product_id:
        raise CommandRejectedError("Product ID is required")
    if cmd.quantity < 0:
        raise CommandRejectedError("Quantity cannot be negative")
    if cmd.low_stock_threshold < 0:
        raise CommandRejectedError("Low stock threshold cannot be negative")

    log.info("initializing_stock", product_id=cmd.product_id, quantity=cmd.quantity)

    event = domains.StockInitialized(
        product_id=cmd.product_id,
        quantity=cmd.quantity,
        low_stock_threshold=cmd.low_stock_threshold,
        initialized_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_receive_stock(command_book, command_any, state: InventoryState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Inventory not initialized")

    cmd = domains.ReceiveStock()
    command_any.Unpack(cmd)

    if cmd.quantity <= 0:
        raise CommandRejectedError("Quantity must be positive")

    log.info("receiving_stock", quantity=cmd.quantity, reference=cmd.reference)

    event = domains.StockReceived(
        quantity=cmd.quantity,
        new_on_hand=state.on_hand + cmd.quantity,
        reference=cmd.reference,
        received_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_reserve_stock(command_book, command_any, state: InventoryState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Inventory not initialized")

    cmd = domains.ReserveStock()
    command_any.Unpack(cmd)

    if cmd.quantity <= 0:
        raise CommandRejectedError("Quantity must be positive")
    if not cmd.order_id:
        raise CommandRejectedError("Order ID is required")
    if cmd.order_id in state.reservations:
        raise CommandRejectedError("Reservation already exists for this order")
    if state.available() < cmd.quantity:
        raise CommandRejectedError(f"Insufficient stock: available {state.available()}, requested {cmd.quantity}")

    log.info("reserving_stock", quantity=cmd.quantity, order_id=cmd.order_id)

    new_available = state.available() - cmd.quantity

    pages = []

    reserved_event = domains.StockReserved(
        quantity=cmd.quantity,
        order_id=cmd.order_id,
        new_available=new_available,
        reserved_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )
    reserved_any = Any()
    reserved_any.Pack(reserved_event, type_url_prefix="type.examples/")
    pages.append(angzarr.EventPage(num=seq, event=reserved_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp()))))

    # Check for low stock alert
    if new_available < state.low_stock_threshold and state.available() >= state.low_stock_threshold:
        alert_event = domains.LowStockAlert(
            product_id=state.product_id,
            available=new_available,
            threshold=state.low_stock_threshold,
            alerted_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
        )
        alert_any = Any()
        alert_any.Pack(alert_event, type_url_prefix="type.examples/")
        pages.append(angzarr.EventPage(num=seq + 1, event=alert_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp()))))

    return angzarr.EventBook(cover=command_book.cover, pages=pages)


def handle_release_reservation(command_book, command_any, state: InventoryState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Inventory not initialized")

    cmd = domains.ReleaseReservation()
    command_any.Unpack(cmd)

    if not cmd.order_id:
        raise CommandRejectedError("Order ID is required")
    if cmd.order_id not in state.reservations:
        raise CommandRejectedError("No reservation found for this order")

    qty = state.reservations[cmd.order_id]

    log.info("releasing_reservation", order_id=cmd.order_id, quantity=qty)

    event = domains.ReservationReleased(
        order_id=cmd.order_id,
        quantity=qty,
        new_available=state.available() + qty,
        released_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_commit_reservation(command_book, command_any, state: InventoryState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Inventory not initialized")

    cmd = domains.CommitReservation()
    command_any.Unpack(cmd)

    if not cmd.order_id:
        raise CommandRejectedError("Order ID is required")
    if cmd.order_id not in state.reservations:
        raise CommandRejectedError("No reservation found for this order")

    qty = state.reservations[cmd.order_id]

    log.info("committing_reservation", order_id=cmd.order_id, quantity=qty)

    event = domains.ReservationCommitted(
        order_id=cmd.order_id,
        quantity=qty,
        new_on_hand=state.on_hand - qty,
        committed_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )

File: examples/ruby/inventory/lib/inventory_logic.rb
# frozen_string_literal: true

require 'time'
require_relative 'command_validation_error'

# Pure business logic for inventory aggregate.
# No gRPC dependencies - can be tested in isolation.
module InventoryLogic
  Reservation = Struct.new(:order_id, :quantity, keyword_init: true)

  InventoryState = Struct.new(:sku, :on_hand, :reserved, :reservations, :low_stock_threshold, keyword_init: true) do
    def initialize(sku: '', on_hand: 0, reserved: 0, reservations: [], low_stock_threshold: 10)
      super
    end

    def exists?
      !sku.empty?
    end

    def available
      on_hand - reserved
    end
  end

  LOW_STOCK_THRESHOLD = 10

  class << self
    def rebuild_state(event_book)
      state = InventoryState.new

      return state if event_book.nil? || event_book.pages.empty?

      # Apply events
      event_book.pages.each do |page|
        next unless page.event

        state = apply_event(state, page.event)
      end

      state
    end

    def handle_initialize_stock(state, sku, initial_quantity, low_stock_threshold)
      raise CommandValidationError.failed_precondition('Inventory already exists') if state.exists?
      raise CommandValidationError.invalid_argument('SKU is required') if sku.nil? || sku.empty?
      raise CommandValidationError.invalid_argument('Quantity must be non-negative') if initial_quantity < 0

      Examples::StockInitialized.new(
        sku: sku,
        initial_quantity: initial_quantity,
        low_stock_threshold: low_stock_threshold || LOW_STOCK_THRESHOLD
      )
    end

    def handle_receive_stock(state, quantity, reference)
      raise CommandValidationError.failed_precondition('Inventory does not exist') unless state.exists?
      raise CommandValidationError.invalid_argument('Quantity must be positive') if quantity <= 0

      Examples::StockReceived.new(
        quantity: quantity,
        new_on_hand: state.on_hand + quantity,
        reference: reference || ''
      )
    end

    def handle_reserve_stock(state, order_id, quantity)
      raise CommandValidationError.failed_precondition('Inventory does not exist') unless state.exists?
      raise CommandValidationError.invalid_argument('Order ID is required') if order_id.nil? || order_id.empty?
      raise CommandValidationError.invalid_argument('Quantity must be positive') if quantity <= 0

      if quantity > state.available
        raise CommandValidationError.failed_precondition(
          "Insufficient stock: available #{state.available}, requested #{quantity}"
        )
      end

      existing = state.reservations.find { |r| r.order_id == order_id }
      raise CommandValidationError.failed_precondition("Reservation already exists for order #{order_id}") if existing

      events = []
      events << Examples::StockReserved.new(
        order_id: order_id,
        quantity: quantity,
        new_available: state.available - quantity
      )

      new_available = state.available - quantity
      if new_available <= state.low_stock_threshold && state.available > state.low_stock_threshold
        events << Examples::LowStockAlert.new(
          sku: state.sku,
          current_available: new_available,
          threshold: state.low_stock_threshold
        )
      end

      events
    end

    def handle_release_reservation(state, order_id)
      raise CommandValidationError.failed_precondition('Inventory does not exist') unless state.exists?
      raise CommandValidationError.invalid_argument('Order ID is required') if order_id.nil? || order_id.empty?

      reservation = state.reservations.find { |r| r.order_id == order_id }
      raise CommandValidationError.failed_precondition("No reservation found for order #{order_id}") unless reservation

      Examples::ReservationReleased.new(
        order_id: order_id,
        quantity: reservation.quantity,
        new_available: state.available + reservation.quantity
      )
    end

    def handle_commit_reservation(state, order_id)
      raise CommandValidationError.failed_precondition('Inventory does not exist') unless state.exists?
      raise CommandValidationError.invalid_argument('Order ID is required') if order_id.nil? || order_id.empty?

      reservation = state.reservations.find { |r| r.order_id == order_id }
      raise CommandValidationError.failed_precondition("No reservation found for order #{order_id}") unless reservation

      Examples::ReservationCommitted.new(
        order_id: order_id,
        quantity: reservation.quantity,
        new_on_hand: state.on_hand - reservation.quantity
      )
    end

    private

    def apply_event(state, event_any)
      type_url = event_any.type_url

      if type_url.end_with?('StockInitialized')
        event = Examples::StockInitialized.decode(event_any.value)
        InventoryState.new(
          sku: event.sku,
          on_hand: event.initial_quantity,
          reserved: 0,
          reservations: [],
          low_stock_threshold: event.low_stock_threshold
        )
      elsif type_url.end_with?('StockReceived')
        event = Examples::StockReceived.decode(event_any.value)
        InventoryState.new(
          sku: state.sku,
          on_hand: event.new_on_hand,
          reserved: state.reserved,
          reservations: state.reservations,
          low_stock_threshold: state.low_stock_threshold
        )
      elsif type_url.end_with?('StockReserved')
        event = Examples::StockReserved.decode(event_any.value)
        new_reservations = state.reservations + [Reservation.new(order_id: event.order_id, quantity: event.quantity)]
        InventoryState.new(
          sku: state.sku,
          on_hand: state.on_hand,
          reserved: state.reserved + event.quantity,
          reservations: new_reservations,
          low_stock_threshold: state.low_stock_threshold
        )
      elsif type_url.end_with?('ReservationReleased')
        event = Examples::ReservationReleased.decode(event_any.value)
        new_reservations = state.reservations.reject { |r| r.order_id == event.order_id }
        InventoryState.new(
          sku: state.sku,
          on_hand: state.on_hand,
          reserved: state.reserved - event.quantity,
          reservations: new_reservations,
          low_stock_threshold: state.low_stock_threshold
        )
      elsif type_url.end_with?('ReservationCommitted')
        event = Examples::ReservationCommitted.decode(event_any.value)
        new_reservations = state.reservations.reject { |r| r.order_id == event.order_id }
        InventoryState.new(
          sku: state.sku,
          on_hand: event.new_on_hand,
          reserved: state.reserved - event.quantity,
          reservations: new_reservations,
          low_stock_threshold: state.low_stock_threshold
        )
      else
        state
      end
    end
  end
end
